scripts/ci-deploy.sh

```diff
@@ -21,29 +22,48 @@ if [[ -z ${ANSIBLE_FOLDERS} ]]; then
   ANSIBLE_FOLDERS="infrastructure"
 fi
 
+if [[ -d ${cdk_app_dir} ]]; then
+  INFRASTRUCTURE_FOLDERS=${cdk_app_dir}
+elif [[ -z ${INFRASTRUCTURE_FOLDERS} ]]; then
+  INFRASTRUCTURE_FOLDERS=${ANSIBLE_FOLDERS}
+fi
+
 # shellcheck source=.
 . "$dir"/ci-include.sh
 
-pip --quiet --disable-pip-version-check --no-color install ansible boto3 requests pyyaml awscli netaddr
-
-# install the collection for CI/CD
-ansible-galaxy collection install --force git+https://github.com/ringier-data/ops-ci-aws.git,main
-
-MODULES=("${ANSIBLE_FOLDERS//,/ }")
+MODULES=("${INFRASTRUCTURE_FOLDERS//,/ }")
 
 for module in "${MODULES[@]}"; do
   echo Deploying "${module}" module...
   pushd "${module}"
 
-  if [[ -f "requirements.txt" ]]; then
-    pip install -r requirements.txt
-  fi
+  if [[ -f "./package.json" && -f "./package-lock.json" ]]; then
+    echo "CDK app detected..."
 
-  if [[ -f "playbook.yml" ]]; then
-    ansible-playbook -e env="$ENV" -v playbook.yml
+    echo "Installing dependencies..."
+    npm --no-color ci
+
+    npm --no-color run cdk diff -- --context env=${ENV}
+
+    echo "Deploying to env=${ENV}..."
+    npm --no-color run cdk deploy --all --require-approval=never -- --context env=${ENV}
   else
-    echo "Could not find playbook.yml!"
-    exit 1
+    echo "Deploying Ansible module..."
+    pip --quiet --disable-pip-version-check --no-color install ansible boto3 requests pyyaml awscli netaddr
+
+    # install the collection for CI/CD
+    ansible-galaxy collection install --force git+https://github.com/ringier-data/ops-ci-aws.git,main
+
+    if [[ -f "requirements.txt" ]]; then
+      pip install -r requirements.txt
+    fi
+
+    if [[ -f "playbook.yml" ]]; then
+      ansible-playbook -e env="$ENV" -v playbook.yml
+    else
+      echo "Could not find playbook.yml!"
```

Comment:
Revised this so that the check for playbook.yml happens on the same level as for package.json/package-lock.json. A generic print is done if neither CDK app or Ansible playbook is found

Reply:
Yes, I agree.

Pointing the CI to a wrong folder could be treated like a "syntax error", which suggests a potential but fatal error. It should fail as early as possible, ideally before doing the real thing. `exit 1` for each conditional branch will stop the entire CI/CD flow immediately, this is much better than letting the code flow wildly run into nowhere. But :) the complexity comes from the details when we think about the edge cases. e.g. I have `INFRASTRUCTURE_FOLDERS=storage,api`. `storage` deployed successfully, but `api` is just an empty folder. The build will fail because the the loop step at `api` returned non-zero. It means we ended up with a partial deployment which might lead to some difficult-to-recover state.

A slightly better approach might be to loop it once firstly, like a dry-run, checking that we have ((`package.json` AND `package-lock.json`) OR `playbook.yml`) in EACH folder specified by INFRASTRUCTURE_FOLDERS. Then loop again to do the heavy-lifting work. It improves the case, but reduces the code readability.

I leave it to you to decide. Either way is okay for me.

Reply:
Agree with the tradeoffs. Now that were onto it, I implemented the dry-run approach and it doesn't look too bad in the code so I think we might as well include it.

Reply:
Agree with the tradeoffs. Now that we're onto it, I implemented the dry-run approach and it doesn't look too bad in the code. So I would say we can have that extra check in place